Assets/Countly/Plugins/CountlySDK/Services/UserDetailsCountlyService.cs

```diff
@@ -99,127 +232,111 @@ public async Task SetUserDetailsAsync(CountlyUserDetailsModel userDetailsModel)
-            Log.Info("[UserDetailsCountlyService] SetCustomUserDetails " + (customDetail != null));
-
             if (!_consentService.CheckConsentInternal(Consents.Users)) {
+                Log.Debug("[UserDetailsCountlyService] AddCustomDetailToRequestQueue, consent is not given, ignoring the request.");
                 return;
             }
 
-            if (customDetail == null || customDetail.Count == 0) {
-                Log.Warning("[UserDetailsCountlyService] SetCustomUserDetails : Provided custom detail 'customDetail' can't be null or empty.");
-
+            if (segments == null || segments.Count == 0) {
+                Log.Warning("[UserDetailsCountlyService] AddCustomDetailToRequestQueue, Provided custom detail 'customDetail' can't be null or empty.");
                 return;
             }
 
-            AddCustomDetailToRequestQueue(customDetail);
+            IDictionary<string, object> customDetail = FixSegmentKeysAndValues(segments);
+
+            Dictionary<string, object> requestParams =
+                new Dictionary<string, object>
+                {
+                    { "user_details",
+                        JsonConvert.SerializeObject(
+                            new Dictionary<string, object>
+                            {
+                                { "custom", customDetail }
+                            })
+                    }
+                };
+            _requestCountlyHelper.AddToRequestQueue(requestParams);
+            _ = _requestCountlyHelper.ProcessQueue();
         }
 
         /// <summary>
         /// Send provided values to server.
         /// </summary>
-        /// <returns></returns>
-        public async Task SaveAsync()
+        private async Task SaveAsyncInternal()
         {
+            if (!_consentService.CheckConsentInternal(Consents.Users)) {
+                Log.Debug("[UserDetailsCountlyService] SaveAsyncInternal, consent is not given, ignoring the request.");
+                return;
+            }
+
             lock (LockObj) {
                 if (!CustomDataProperties.Any()) {
                     return;
                 }
 
-                Log.Info("[UserDetailsCountlyService] SaveAsync");
-
-
+                Log.Info("[UserDetailsCountlyService] SaveAsyncInternal");
                 CountlyUserDetailsModel model = new CountlyUserDetailsModel(CustomDataProperties);
-
                 CustomDataProperties = new Dictionary<string, object> { };
                 AddCustomDetailToRequestQueue(CustomDataProperties);
             }
+
             await Task.CompletedTask;
         }
 
-
         /// <summary>
         /// Sets custom provide key/value as custom property.
         /// </summary>
         /// <param name="key">string with key for the property</param>
         /// <param name="value">string with value for the property</param>
-        public void Set(string key, string value)
+        private void SetInternal(string key, string value, bool setOnce)
```

Comment:
but setOnce is a modifier, its context is not related to "Set" I feel. 

Maybe you could create a ModifyInternal and provide only modifier key to it.